src/main/java/org/truffleruby/core/method/MethodNodes.java

```diff
@@ -275,16 +291,15 @@ public DynamicObject methodUnimplement(DynamicObject rubyMethod) {
 
     private static class SetReceiverNode extends RubyNode {
 
-        private final Object receiver;
         @Child private DirectCallNode methodCallNode;
 
-        public SetReceiverNode(Object receiver, CallTarget methodCallTarget) {
-            this.receiver = receiver;
+        public SetReceiverNode(CallTarget methodCallTarget) {
             this.methodCallNode = DirectCallNode.create(methodCallTarget);
         }
 
         @Override
         public Object execute(VirtualFrame frame) {
+            Object receiver = RubyArguments.getSelf(RubyArguments.getDeclarationFrame(frame));
             RubyArguments.setSelf(frame, receiver);
```

Comment:
Actually, is it needed to set `self` here or would it be enough to set the `self` field of the proc, since that's what's used when calling a Proc?
Maybe it's a problem for `instance_exec(&method(:foo))`.